feat(volume-events-controller): add controller to send volume events to configured server

mittachaitu opened the pull request:
**What this PR does:**
This PR adds a feature(volume-events-controller) to send volume
events(create & delete) to configured remote server. volume-event-controller
will make sure to send events only if `events.openebs.io/required: true` annotation
was set on PV and then remove finalizers from resources.

**Changes in PR**:
- Adds support to send nfs volume created information to the remote server.
- Adds support to send nfs volume deleted information to the remote server.
- Adds UnitTest, YAMLs and Makefile changes.

**Configuration**:
As of now server endpoint, authentication token, and NFS server namespace
are configured via the following environment variables on volume-events-exporter
sidecar container. For more detailed YAML take a look at deploy/kubectl/openebs-nfs-provisioner.yaml    
```yaml
...
...
...
    - name: CALLBACK_URL
       value: "http://192.168.0.5:9000/event-server"
    - name: CALLBACK_TOKEN
       value: "yJhbGciOiJIUzI1NiIsInR5cCI6I"
    - name: OPENEBS_IO_NFS_SERVER_NS
       value: "openebs"
```

**Note**:
- Finalizers shouldn't be removed manually by users. If they were
  removed then `volume-event-controller` can't ensure to send metrics.
- As of now only support to send nfs volume information and it can be easily
   extended for other volumes.
 
**Sample Screenshot**:
- Following screen shot represents the data sent to server by volume-event-controller side car
![Screenshot from 2021-07-28 19-48-16](https://user-images.githubusercontent.com/13776912/127522576-b483061a-c85e-4891-9273-e180a77a70f2.png)
- Sample JSON format received at server side.
[output.txt](https://github.com/mayadata-io/volume-events-exporter/files/6901778/output.txt)

**Reference Document**:
- [Design document](https://docs.google.com/document/d/1YxJCmZLrxxG3mqBQX924waZYh449RFMBiKVDiJuamnk/edit?usp=sharing)


 Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

Commit:
feat(events-controller): add controller to send volume events to configured server

This commit add a feature(volume-event-controller) to send volume
events(create & delete) to configured remote server. volume-event-controller
will make sure to send events and then remove finalizers from resource.

Configuration:
As of now server endpoint, authentication token and NFS server namespace
are configured via following environment variables:

```yaml
...
...
...
  - name: CALLBACK_URL
    value: "http://192.168.0.5:9000/event-server"
  - name: CALLBACK_TOKEN
    value: "yJhbGciOiJIUzI1NiIsInR5cCI6I"
  - name: OPENEBS_IO_NFS_SERVER_NS
    value: "openebs"
```

Note:
- Finalizers shouldn't be removed manually by users. If they were
  removed then volume-event-controller can't ensure to send metrics.

Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

Commit:
This commit adds UnitTest for the code base

Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

Commit:
This commit adds required artifacts for deploying volume-events-exporter

Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

Commit:
This commit adds github workflows

Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

mittachaitu commented on pkg/controller/pvcontroller/sync.go:
```diff
@@ -0,0 +1,195 @@
+	return pvObj.DeletionTimestamp == nil && isCreateVolumeEventSent(pvObj)
+}
+
+// isCreateVolumeEventSent will return true if volume has
+// suffix(event.openebs.io/volume-create) in annotations and value is sent
+func isCreateVolumeEventSent(pvObj *corev1.PersistentVolume) bool {
+	for key, value := range pvObj.Annotations {
+		if strings.HasSuffix(key, collectorinterface.OpenebsCreateAnnotationSuffix) {
+			return value == collectorinterface.OpenebsSentAnnotationValue
+		}
+	}
+	return false
+}
+
+// isDeleteVolumeEventSent will return true if volume has
+// suffix(event.openebs.io/volume-delete) in annotations and value is sent
+func isDeleteVolumeEventSent(pvObj *corev1.PersistentVolume) bool {
+	for key, value := range pvObj.Annotations {
+		if strings.HasSuffix(key, collectorinterface.OpenebsDeleteAnnotationSuffix) {
+			return value == collectorinterface.OpenebsSentAnnotationValue
+		}
+	}
+	return false
+}
```
I think I should move these predicates to collector-interface(Which will make more sense).

mynktl replied:
we can keep it here also, since we are checking for both, create and delete, conditions on the underlying object, PV

mittachaitu commented on pkg/controller/pvcontroller/sync.go:
```diff
@@ -0,0 +1,195 @@
+			break
+		}
+	}
+	if eventFinalizer == "" {
+		klog.Warningf("unable to find volume %s type from finalizers %v", pvObj.Name, pvObj.Finalizers)
+		return nil, nil
+	}
```
Currently, we are sending events only if there is suffix `events.openebs.io/finalizer` in any one of the finalizers. I don't think this check is no longer required(since we are identifying volume type based on label[we can also do it by looking at provisioned name in PV annotation]). After getting the sender interface rest of the code flow should take care instead of checking here(If create/delete event is not yet pushed sent them).

mittachaitu replied:
Removed this check as it is not necessary after going with annotation `events.openebs.io/required: true`

mynktl commented on cmd/volume-events-exporter/app/start.go:
```diff
@@ -0,0 +1,149 @@
+/*
```
how about moving this file to `pkg/cli`?

mittachaitu replied:
hmm.. Got it in the future if we want to do it a bit differently we can easily extend it.

mittachaitu replied:
Moved this file to `pkg/cmd/` as per offline discussion

mynktl commented on cmd/volume-events-exporter/app/start.go:
```diff
@@ -0,0 +1,149 @@
+
+	run := func(context.Context) {
+		var wg sync.WaitGroup
+		wg.Add(3)
```
better to add before starting go routine

mittachaitu replied:
Added wait just before starting Go routine

mynktl commented on cmd/volume-events-exporter/app/start.go:
```diff
@@ -0,0 +1,149 @@
+		go func() {
+			pvInformer.Informer().Run(stopCh)
+			wg.Done()
+		}()
```
why are we running informers separately? we can start the root informer `kubeInformerFactory`.

mittachaitu replied:
No specific reason, I will make it a single call

mittachaitu replied:
Updated to use `kubeInformerFactory`

mynktl commented on pkg/collectorinterface/collector_interface.go:
```diff
@@ -0,0 +1,49 @@
+
+type VolumeEventCollector interface {
+	// CollectCreateEventData should return data required for create event
+	CollectCreateEventData() (string, error)
```
better to rename it to `CollectCreateEvents` and others also

mittachaitu replied:
Done

mynktl commented on pkg/collectorinterface/collector_interface.go:
```diff
@@ -0,0 +1,49 @@
+	// CollectDeleteEventData should return data required for delete event
+	CollectDeleteEventData() (string, error)
+	// RemoveEventFinalizer should remove the finalizer on all dependent resources
+	RemoveEventFinalizer() error
```
how about naming it to `RemoveFinalizer`?

mittachaitu replied:
To me, just RemoveFinalizer is more generic and not giving enough context unless someone knows what this interface is doing(Like `CollectCreateEvents` this `RemoveEventsFinalizer/RemoveEventFinalizer` is conveying the meaning).

mynktl commented on pkg/collectorinterface/const.go:
```diff
@@ -0,0 +1,24 @@
+package collectorinterface
+
+const (
+	OpenebsCreateAnnotationSuffix = "event.openebs.io/volume-create"
```
`OpenebsCreateAnnotationSuffix` -> `AnnotationVolumeCreateEvent`, same for delete one.

mittachaitu replied:
How about `VolumeCreateEventAnnotationSuffix`? Since it is not an entire event annotation name I want to convey it is a suffix part of annotation and it can have prefix as `nfs.<event.openebs.io/volume-create>`/`local.event.openebs.io/volume-create`.

mittachaitu replied:
Updated to
`OpenebsCreateAnnotationSuffix` --> `VolumeCreateEventAnnotation`  &
`OpenebsDeleteAnnotationSuffix` --> `VolumeDeleteEventAnnotation`

mynktl commented on pkg/collectorinterface/collector_interface.go:
```diff
@@ -0,0 +1,49 @@
+
+type DataType string
+
+const (
```
move this to pkg/collectorinterface/const.go

mittachaitu replied:
Done

mynktl commented on pkg/collectorinterface/tokenauth/send.go:
```diff
@@ -0,0 +1,97 @@
+const (
+	// postMethod is used to send http POST request
+	postMethod   = "POST"
+	okStatusCode = 200
```
no need to define a new constant. we can use `http.StatusOK`

mittachaitu replied:
I missed it... :+1:

mittachaitu replied:
Re-used `http.StatusOK`

mynktl commented on pkg/collectorinterface/tokenauth/send.go:
```diff
@@ -0,0 +1,97 @@
+	return &TokenClient{
+		serverURL:            env.GetCallBackServerURL(),
+		serverAuthToken:      env.GetCallBackServerAuthToken(),
+		dataType:             collectorinterface.JSONDataType,
```
let caller pass this value

mynktl replied:
or we can fetch it through `collectorInterface` since backend PV engine will be sending data in serialized json/yaml format.

mittachaitu replied:
> we can fetch it through collectorInterface since backend PV engine will be sending data in serialized json/yaml format

Do you mean by adding a different function in `collectorInterface`? Something like caller will look like `obj.Type(json/yaml).CollectCreateEvents()`?

mittachaitu replied:
Updated as per discussion

mynktl commented on pkg/controller/pvcontroller/controller.go:
```diff
@@ -0,0 +1,156 @@
+/*
```
filepath : `pkg/controller/pvcontroller/controller.go` -> `pkg/controller/pvcontroller.go`

mittachaitu replied:
I took one layer down so that we can be extended to `n` controllers in the same repo but I don't see in near feature with that requirement so we can remove extra package `pvcontroller`

mittachaitu replied:
Followed same pattern as per discussion...
```sh
 tree pkg/controller/
pkg/controller/
├── base_controller.go
├── pv_controller.go
├── pv_controller_sync.go
├── pv_controller_sync_test.go
└── types.go                                                       ## Contains interface to run controllers

0 directories, 5 files
```

mynktl commented on pkg/controller/pvcontroller/controller.go:
```diff
@@ -0,0 +1,156 @@
+)
+
+var (
+	controllerAgentName = "volume-events-controller"
```
`controllerAgentName` -> `volumeEventControllerName`

mittachaitu replied:
Updated to `volumeEventControllerName`

mynktl commented on pkg/controller/pvcontroller/controller.go:
```diff
@@ -0,0 +1,156 @@
+
+// ClientsetListers contains clientset to perform write operations on
+// K8s resources and listers to get/list from shared informer's store
+type ClientsetListers struct {
```
this can be merged with `PVMetricsController`

mittachaitu replied:
Done

mynktl commented on pkg/controller/pvcontroller/controller.go:
```diff
@@ -0,0 +1,156 @@
+
+// PVMetricsControllerConfig helps to create new instance
+// of PVMetricsController
+type PVMetricsControllerConfig struct {
```
since arguments are few we can avoid defining struct for this

mynktl replied:
nit: `PVMetricsController` -> `PvEventController`

mynktl replied:
or `volumeEventController`

mittachaitu replied:
I will go with `PVEventController`

mittachaitu replied:
Updated to codebase to use `PVEventController`

mynktl commented on pkg/controller/pvcontroller/controller.go:
```diff
@@ -0,0 +1,156 @@
+		},
+		workqueue:          workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerAgentName),
+		NFSServerNamespace: env.GetNFSServerNamespace(),
+		pvcListerSynced:    config.PVCInformer.Informer().HasSynced,
```
we can wait for informers to sync before running the controller. That way we can avoid tracking `HasSynced`.

mynktl commented on pkg/controller/pvcontroller/controller.go:
```diff
@@ -0,0 +1,156 @@
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(pv); err != nil {
+		runtime.HandleError(err)
```
lets print/dump object here

mittachaitu replied:
Re-Used same code as specified [here](https://github.com/mayadata-io/dmaas-operator/blob/467131e06d8109d2af1cd81df26cfda067084700/pkg/controller/base_controller.go#L179-L188)

mynktl commented on pkg/controller/pvcontroller/runner.go:
```diff
@@ -0,0 +1,107 @@
+	klog.InfoS("starting up  callback volume metrics controller")
+	defer klog.InfoS("shutting down callback volume metrics controller")
+
+	if !cache.WaitForNamedCacheSync(controllerAgentName, stopCh, pController.pvListerSynced, pController.pvcListerSynced) {
```
this can be moved to the caller or we can add an array with all required informers  under `pController`.

mynktl replied:
how about creating a separate interface to run/create the controller.

mittachaitu replied:
> how about creating a separate interface to run/create the controller.

IMO we can go with the interface option if we want either controller1 or controller2. In this case, I am seeing only one controller at the moment so thinking it is, for now, we can go without an interface.  I will this as an array with all required informers.

mittachaitu replied:
Pushed into array with all required informers and waiting for them to sync before starting controller

mynktl commented on pkg/controller/pvcontroller/sync.go:
```diff
@@ -0,0 +1,195 @@
+)
+
+const (
+	eventRequiredAnnotationKey   = "events.openebs.io/required"
```
nit: `eventRequiredAnnotationKey` -> `AnnotationProcessEventKey`

mynktl commented on pkg/controller/pvcontroller/sync.go:
```diff
@@ -0,0 +1,195 @@
+// syncToSendVolumeEvents reconciles PersistentVolume and will
+// send volume information to configured callback URL if it is required
+// send (or) volume event information is not yet sent
+func (pController *PVMetricsController) syncToSendVolumeEvents(key string) error {
```
nit: `syncToSendVolumeEvents` -> `processVolumeEvents`

mittachaitu replied:
Done

mynktl commented on pkg/controller/pvcontroller/sync.go:
```diff
@@ -0,0 +1,195 @@
+	}
+
+	// Send create information
+	if !isCreateVolumeEventSent(pvObj) {
```
let's move this (also for delete) to a separate function

mittachaitu replied:
Moved into separate functions like `sendCreateEvent` & `sendDeleteEvent`

mynktl commented on pkg/controller/pvcontroller/sync.go:
```diff
@@ -0,0 +1,195 @@
+// 1. Retrun true if volume doesn't require event to be exported(based on annotation "events.openebs.io/required": "true")
+// 2. Retrun true if deletion timestamp is not set and create event already sent
+// 3. else return false
+func canSkipEventCollection(pvObj *corev1.PersistentVolume) bool {
```
nit: shouldSendEvent

mittachaitu replied:
Updated to shouldSendEvent

mynktl commented on pkg/controller/pvcontroller/sync.go:
```diff
@@ -0,0 +1,195 @@
+
+// isCreateVolumeEventSent will return true if volume has
+// suffix(event.openebs.io/volume-create) in annotations and value is sent
+func isCreateVolumeEventSent(pvObj *corev1.PersistentVolume) bool {
```
nit: `volumeCreateEventSent`

mittachaitu replied:
I felt isCreateVolumeEventSent is a long name but it is conveying purpose so kept as is in code base

mynktl commented on pkg/env/env.go:
```diff
@@ -0,0 +1,53 @@
+
+var (
+	// NFSServerNamespace defines the namespace of NFS Server resources
+	NFSServerNamespace = "OPENEBS_IO_NFS_SERVER_NS"
```
lets use `openebs/dynamic-nfs-provisioner`  package for nfs-server related config params

mittachaitu replied:
Thought about it(easy for maintenance) but avoided it since we have to do the import again.

mittachaitu replied:
Updating to use `openebs/dynamic-nfs-provisioner` is causing other issues since volume-event-exporter is using v0.21.3 version whereas `openebs/maya`(indirect dependency via `openebs/dynamic-nfs-provisioner`) is using old version v0.17.3. So I didn't update this like will take this change later in the future.

mynktl replied:
ok

mynktl commented on pkg/nfspv/nfsvolume.go:
```diff
@@ -0,0 +1,302 @@
+		volumeData.NFSPVC.DeletionGracePeriodSeconds = nil
+	}
+
+	volumeData.NFSPV.DeletionTimestamp = nil
```
nit: better to add a comment here stating, to send create and delete event, if create event missed, we are setting this values to nil....

mittachaitu replied:
Added comment before starting the function

mynktl commented on pkg/signals/signals.go:
```diff
@@ -0,0 +1,42 @@
+	go func() {
+		<-c
+		close(stop)
+		<-c
```
not required
```suggestion
```

mittachaitu replied:
Removed

Commit:
This commit address review comments

Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

Commit:
This commit fixes volume-event-exporter.Dockerfile dockerfile

Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

Commit:
This commit updates codebase to identify nfs volume based on openebs.io/cas-type label and also updates the build files

Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

mynktl commented on buildscripts/test.sh:
```diff
@@ -0,0 +1,42 @@
+#  exit 0
+#fi
+
+echo "" > coverage.txt
```
why do we need to create an empty file here?

mittachaitu replied:
Not required

mynktl commented on buildscripts/volume-events-exporter/volume-event-exporter.Dockerfile:
```diff
@@ -0,0 +1,57 @@
+# Copyright 2021 The MayaData Authors
```
let's remove this file for now

mittachaitu replied:
any specific reason to remove it?

mynktl replied:
we are not using it.

mittachaitu replied:
We are using it during build time, to push [multi-arch images](https://github.com/mayadata-io/volume-events-exporter/pull/1/files#diff-5c3fa597431eda03ac3339ae6bf7f05e1a50d6fc7333679ec38e21b337cb6721R125-R133) to mayadataio docker repo.

mynktl commented on deploy/kubectl/openebs-nfs-provisioner.yaml:
```diff
@@ -0,0 +1,210 @@
+        #  value: "http://173.168.20.5:9000/event-server"
+        #  CALLBACK_TOKEN defines the authentication token required to interact with server.
+        #- name: CALLBACK_TOKEN
+        #  value: "eyJhbGciOiJIUzI1NiIsI"
```
nit:
```suggestion
        # CALLBACK_TOKEN defines the authentication token required to interact with server.
        #- name: CALLBACK_TOKEN
        #  value: "eyJhbGciOiJIUzI1NiIsI"
```

mittachaitu replied:
Done

mynktl commented on Makefile:
```diff
@@ -0,0 +1,200 @@
+clean:
+	@echo "--> Cleaning Directory"
+	go clean -testcache
+	rm -rf bin/
```
we are copying binary into ${GOPATH}/bin also. let's remove those file as part of `clean`

mittachaitu replied:
We are pushing `golangci-lint` into ${GOPATH}/bin... Which was copied as part of the bootstrap process. Do we need to remove them?

mynktl replied:
not required.

mynktl commented on Makefile:
```diff
@@ -0,0 +1,200 @@
+# during build
+.PHONY: bootstrap
+bootstrap: install-golangci-lint
+	@echo "Currently no tools required"
```
we are installing golangci-lint tool as part of bootstrap. Can we move this under bootstrap instead of having the separate build target?

mittachaitu replied:
Moved `install-golangci-lint` tool as part of bootstrap instead of separate target

mynktl commented on pkg/cmd/events_exporter.go:
```diff
@@ -0,0 +1,133 @@
+// StartVolumeEventsController will start volume events collector controller
+func StartVolumeEventsController() error {
+	klog.InitFlags(nil)
+	err := flag.Set("logtostderr", "true")
```
This will make `-logtostderr` arg unusable. By default, Klog print logs to stderr only. We can avoid setting this value here.

mittachaitu replied:
Removed setting `logtostderr` to true

mynktl commented on pkg/cmd/events_exporter.go:
```diff
@@ -0,0 +1,133 @@
+		// Start registered informers
+		go func() {
+			kubeInformerFactory.Start(stopCh)
+			wg.Done()
```
we don't need to run this in go routine.

mittachaitu replied:
Done

mynktl commented on pkg/controller/pv_controller.go:
```diff
@@ -0,0 +1,143 @@
+	pvcInformer corev1informer.PersistentVolumeClaimInformer,
+	numWorker int) Controller {
+	eventBroadcaster := record.NewBroadcaster()
+	// eventBroadcaster.StartLogging(klog.Infof)
```
lets add logging also

mittachaitu replied:
Enabled recording event based on `-generate-k8s-events` flag value. Defaults to `false` which means events will not get generated, if enabled i.e `--generate-k8s-events=true` then events will be generated on PersistentVolume resource.

mynktl commented on pkg/controller/pv_controller.go:
```diff
@@ -0,0 +1,143 @@
+}
+
+func (pController *PVEventController) updatePV(oldObj, newObj interface{}) {
+	//TODO: Undo below comments if required
```
need to remove it

mittachaitu replied:
Done

mynktl commented on pkg/controller/base_controller.go:
```diff
@@ -0,0 +1,183 @@
+	}
+
+	if c.sync != nil && c.syncPeriod == 0 {
+		return errors.Errorf("syncperiod not set")
```
nit: syncperiod -> sync period

mittachaitu replied:
updated to `sync period`

mynktl commented on pkg/controller/pv_controller_sync.go:
```diff
@@ -0,0 +1,219 @@
+
+	err = pController.sync(pvObj)
+	// TODO: If events needs to be generated undo the comments
+	// if err != nil {
```
lets enable the event

mittachaitu replied:
If we enable the events, then users will get to know we are sending events somewhere. Is it fine?

mittachaitu replied:
Enabled based on `--generate-k8s-event` flag value

mynktl commented on pkg/nfspv/nfsvolume.go:
```diff
@@ -0,0 +1,340 @@
+)
+
+const (
+	OpenEBSCASLabelKey      = "openebs.io/cas-type"
```
this is not specific to nfs PV. let's move it in pkg/controller

mittachaitu replied:
Moved constant to pkg/controller/base_controller.go when the configuration is exploded then we will move to separate file or to openebs/api repository.

mynktl commented on pkg/controller/pv_controller_sync.go:
```diff
@@ -0,0 +1,219 @@
+
+func (pController *PVEventController) getEventSender(pvObj *corev1.PersistentVolume) (collectorinterface.EventsSender, error) {
+	// Add more types based on underlying volume type
+	if value, ok := pvObj.Labels[nfspv.OpenEBSCASLabelKey]; ok && value == nfspv.OpenEBSNFSCASLabelValue {
```
nit: let's use switch case here

mittachaitu replied:
Updated the code... Which can also accumulate CSI based PVs

mynktl commented on pkg/controller/pv_controller_sync.go:
```diff
@@ -0,0 +1,219 @@
+				pController.nfsServerNamespace,
+				collectorinterface.JSONDataType)), nil
+	}
+	return nil, errors.Errorf("event sender is not available for volume %s", pvObj.Name)
```
lets log `value` here, to know the pv type?

mittachaitu replied:
Updated the error message to have castype value also.

mynktl commented on pkg/controller/pv_controller_sync.go:
```diff
@@ -0,0 +1,219 @@
+
+// sendDeleteEvent will push delete volume events to configured server
+// NOTE: If event is already sent then sendDeleteEvent will return nil
+func (pController *PVEventController) sendDeleteEvent(
```
lets add a comment about removing finalizer also

mittachaitu replied:
Done

mynktl commented on pkg/controller/pv_controller.go:
```diff
@@ -0,0 +1,143 @@
+	// workqueue workqueue.RateLimitingInterface
+
+	// nfsServerNamespace states the namespace of NFSServer deployment
+	nfsServerNamespace string
```
let's move this to pkg/nfspv

mynktl commented on pkg/collectorinterface/const.go:
```diff
@@ -0,0 +1,31 @@
+package collectorinterface
+
+const (
+	VolumeCreateEventAnnotation     = "event.openebs.io/volume-create"
```
lets add comment for these constants

mittachaitu replied:
Updated with constants with code comments

mynktl commented on pkg/collectorinterface/tokenauth/send.go:
```diff
@@ -0,0 +1,93 @@
+)
+
+type TokenClient struct {
+	// server url holds the URL to communicate with server
```
nit: 
```suggestion
	// serverURL holds the URL to communicate with server
```

mittachaitu replied:
Done

mynktl commented on pkg/nfspv/nfsvolume.go:
```diff
@@ -0,0 +1,340 @@
+//		 timestamp for create event we are mutating deletion timestamp
+//		 fields with nil value.
+func (n *nfsVolume) CollectCreateEvents() (string, error) {
+
```
nit:
```suggestion
```

mynktl commented on pkg/nfspv/nfsvolume.go:
```diff
@@ -0,0 +1,340 @@
+	backendPVCName := "nfs-" + nfsPV.Name
+	backendPVC, err := n.getPVCCopy(n.nfsServerNamespace, backendPVCName)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get PVC {%s/%s}", n.nfsServerNamespace, backendPVCName)
```
```suggestion
		return nil, errors.Wrapf(err, "failed to get backing PVC {%s/%s}", n.nfsServerNamespace, backendPVCName)
```

mynktl commented on pkg/nfspv/nfsvolume.go:
```diff
@@ -0,0 +1,340 @@
+
+	backendPV, err := n.getPVCopy(backendPVC.Spec.VolumeName)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get PV %s", backendPVC.Spec.VolumeName)
```
```suggestion
		return nil, errors.Wrapf(err, "failed to get backing PV %s", backendPVC.Spec.VolumeName)
```

mittachaitu replied:
Updated the error message

mynktl commented on pkg/nfspv/nfsvolume.go:
```diff
@@ -0,0 +1,340 @@
+}
+
+func (n *nfsVolume) CollectDeleteEvents() (string, error) {
+	// TODO: Is below check required?
```
we can drop this check since the caller is taking care of this.

mittachaitu replied:
Removed check

mynktl commented on pkg/nfspv/nfsvolume.go:
```diff
@@ -0,0 +1,340 @@
+	if err != nil {
+		return nil, err
+	}
+	// Update inmemory reference
```
why are we updating in-memory PV reference?

mittachaitu replied:
Since we triggered the patch on PV for the next update it will return conflict errors. To avoid conflicting errors we are updating inmemory reference.

mynktl replied:
ok

mynktl commented on pkg/nfspv/nfsvolume.go:
```diff
@@ -0,0 +1,340 @@
+	}
+}
+
+func getPatchData(oldObj, newObj interface{}) ([]byte, []byte, error) {
```
how about creating a separate package for `getPatchData`, `removeFinalizer`, and other generic functions?

mittachaitu replied:
Make sense!!

mittachaitu replied:
Moved to helper package i.e `pkg/helper/helper.go`

mynktl commented on pkg/controller/pv_controller_sync.go:
```diff
@@ -0,0 +1,219 @@
+	return true, err
+}
+
+func (pController *PVEventController) sync(pvObj *corev1.PersistentVolume) error {
```
nit: comment

mittachaitu replied:
Added code comments

Commit:
This commit address the review comments

Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

Commit:
This commit adds option via flags to generate Kubernetes events

Adds flag `--generate-k8s-events` to generate Kubernetes events on
corresponding volume

Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

mynktl commented on pkg/controller/pv_controller.go:
```diff
@@ -128,3 +142,27 @@ func GetSyncInterval() time.Duration {
 }
+
+// Event is a wrapper over original Event which will help to generate events
+func (r *Recorder) Event(object runtime.Object, eventtype, reason, message string) {
```
let's move the event code to a separate file.

mittachaitu replied:
Thought of moving when an event has more lines of code... But yeah, good to move before exploded

mittachaitu replied:
Pushed to recorder.go file

Commit:
This commit moves the event generation code to separate file in controller package

Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

Commit:
This commit add license header to recorder.go file

Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>
